File: master_security_python/src/master_security/core/events.py
"""
Master Security Framework - Async Event Bus
=============================================

Event-driven security architecture with async event processing.
Supports real-time threat detection, alerting, and autonomous response.

Features:
    - Async event publishing and subscription
    - Event filtering by type and severity
    - Dead letter queue for failed handlers
    - Event replay for forensic analysis
    - Multi-tenant event isolation

Usage:
    >>> from master_security.core import get_event_bus, SecurityEvent
    >>> bus = get_event_bus()
    >>> bus.subscribe("threat.detected", my_handler)
    >>> bus.publish(SecurityEvent(type="threat.detected", severity="high"))
"""

from __future__ import annotations

import asyncio
import secrets
import threading
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Callable, Optional

from master_security.core.logger import get_logger

logger = get_logger("msf.events")


class EventSeverity(Enum):
    INFO = "info"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class EventType(Enum):
    AUTH_SUCCESS = "auth.success"
    AUTH_FAILURE = "auth.failure"
    THREAT_DETECTED = "threat.detected"
    POLICY_VIOLATION = "policy.violation"
    RATE_LIMIT_HIT = "rate_limit.hit"
    ANOMALY_DETECTED = "anomaly.detected"
    SESSION_HIJACK = "session.hijack"
    CREDENTIAL_STUFFING = "credential.stuffing"
    BRUTE_FORCE = "brute.force"
    XSS_ATTEMPT = "xss.attempt"
    SQLI_ATTEMPT = "sqli.attempt"
    PROMPT_INJECTION = "prompt.injection"
    MALWARE_DETECTED = "malware.detected"
    CONTAINER_ESCAPE = "container.escape"
    DATA_EXFILTRATION = "data.exfiltration"
    AUTONOMOUS_RESPONSE = "autonomous.response"
    CUSTOM = "custom"


@dataclass
class SecurityEvent:
    """
    Security event for the event bus.

    Attributes:
        type: Event type
        severity: Event severity
        source: Event source module
        tenant_id: Multi-tenant identifier
        trace_id: Correlation trace ID
        data: Event payload
        timestamp: Event timestamp
        id: Unique event ID
    """
    type: str
    severity: EventSeverity = EventSeverity.MEDIUM
    source: str = ""
    tenant_id: str = "default"
    trace_id: str = ""
    data: dict[str, Any] = field(default_factory=dict)
    timestamp: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    id: str = field(default_factory=lambda: secrets.token_hex(8))

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "type": self.type,
            "severity": self.severity.value,
            "source": self.source,
            "tenant_id": self.tenant_id,
            "trace_id": self.trace_id,
            "data": self.data,
            "timestamp": self.timestamp,
        }


EventHandler = Callable[[SecurityEvent], Any]
AsyncEventHandler = Callable[[SecurityEvent], Any]


class EventBus:
    """
    Async event bus for security events.

    Thread-safe event publishing and subscription with
    support for sync and async handlers.

    Attributes:
        max_queue_size: Maximum event queue size
        dead_letter_enabled: Enable dead letter queue

    Example:
        >>> bus = EventBus()
        >>> bus.subscribe("threat.detected", handle_threat)
        >>> await bus.publish(SecurityEvent(type="threat.detected"))
    """

    def __init__(
        self,
        max_queue_size: int = 10000,
        dead_letter_enabled: bool = True,
    ) -> None:
        self.max_queue_size = max_queue_size
        self.dead_letter_enabled = dead_letter_enabled
        self._handlers: dict[str, list[EventHandler | AsyncEventHandler]] = {}
        self._dead_letter: list[SecurityEvent] = []
        self._event_history: list[SecurityEvent] = []
        self._max_history = 10000
        self._lock = threading.RLock()
        self._async_lock = asyncio.Lock()
        self._running = False

    def subscribe(
        self,
        event_type: str,
        handler: EventHandler | AsyncEventHandler,
    ) -> None:
        """
        Subscribe to an event type.

        Args:
            event_type: Event type to subscribe to
            handler: Handler function (sync or async)
        """
        with self._lock:
            if event_type not in self._handlers:
                self._handlers[event_type] = []
            self._handlers[event_type].append(handler)
        logger.debug("msf.events.subscribed", event_type=event_type)

    def unsubscribe(
        self,
        event_type: str,
        handler: EventHandler | AsyncEventHandler,
    ) -> None:
        """Unsubscribe from an event type."""
        with self._lock:
            if event_type in self._handlers:
                self._handlers[event_type] = [
                    h for h in self._handlers[event_type] if h != handler
                ]

    def publish_sync(self, event: SecurityEvent) -> None:
        """Synchronously publish a security event."""
        with self._lock:
            self._event_history.append(event)
            if len(self._event_history) > self._max_history:
                self._event_history = self._event_history[-self._max_history:]
            handlers = list(
                self._handlers.get(event.type, [])
                + self._handlers.get("*", [])
            )
        for handler in handlers:
            try:
                if not asyncio.iscoroutinefunction(handler):
                    handler(event)
            except Exception as exc:
                if self.dead_letter_enabled:
                    self._dead_letter.append(event)

    async def publish(self, event: SecurityEvent) -> None:
        """
        Publish a security event to all subscribers.

        Args:
            event: Security event to publish.
        """
        import asyncio as aio

        with self._lock:
            self._event_history.append(event)
            if len(self._event_history) > self._max_history:
                self._event_history = self._event_history[-self._max_history:]

        handlers = []
        with self._lock:
            handlers = list(
                self._handlers.get(event.type, [])
                + self._handlers.get("*", [])
            )

        for handler in handlers:
            try:
                if asyncio.iscoroutinefunction(handler):
                    await handler(event)
                else:
                    handler(event)
            except Exception as exc:
                logger.error("msf.events.handler_error", error=str(exc))
                if self.dead_letter_enabled:
                    self._dead_letter.append(event)

    def get_history(
        self,
        event_type: Optional[str] = None,
        limit: int = 100,
    ) -> list[SecurityEvent]:
        """Get recent event history."""
        with self._lock:
            events = self._event_history
            if event_type:
                events = [e for e in events if e.type == event_type]
            return events[-limit:]

    def get_dead_letter(self) -> list[SecurityEvent]:
        """Get dead letter queue."""
        with self._lock:
            return list(self._dead_letter)


_global_event_bus: EventBus | None = None
_event_bus_lock = threading.Lock()


def get_event_bus() -> EventBus:
    """Get the global event bus instance."""
    global _global_event_bus
    if _global_event_bus is None:
        with _event_bus_lock:
            if _global_event_bus is None:
                _global_event_bus = EventBus()
    return _global_event_bus

File: packages/core/src/core/events.ts
/**
 * Event bus for security event distribution
 * @module core/events
 */

/**
 * Event severity levels
 */
export enum EventSeverity {
  /** Informational event */
  INFO = 'info',
  /** Warning event */
  WARNING = 'warning',
  /** Error event */
  ERROR = 'error',
  /** Critical event */
  CRITICAL = 'critical',
}

/**
 * Event type classification
 */
export enum EventType {
  /** Authentication event */
  AUTHENTICATION = 'authentication',
  /** Authorization event */
  AUTHORIZATION = 'authorization',
  /** Security policy event */
  POLICY = 'policy',
  /** Threat detection event */
  THREAT = 'threat',
  /** System event */
  SYSTEM = 'system',
  /** Audit event */
  AUDIT = 'audit',
  /** Configuration change event */
  CONFIG = 'config',
  /** Rate limiting event */
  RATE_LIMIT = 'rate_limit',
  /** Data access event */
  DATA_ACCESS = 'data_access',
  /** Network event */
  NETWORK = 'network',
}

/**
 * Security event structure
 */
export interface SecurityEvent {
  /** Unique event ID */
  id: string;
  /** Event type */
  type: EventType;
  /** Event severity */
  severity: EventSeverity;
  /** Event timestamp */
  timestamp: Date;
  /** Event source */
  source: string;
  /** Event message */
  message: string;
  /** Event metadata */
  metadata: Record<string, unknown>;
  /** Actor identifier */
  actorId?: string;
  /** Target resource */
  target?: string;
  /** MITRE ATT&CK technique ID */
  mitreTechnique?: string;
  /** Event correlation ID */
  correlationId?: string;
}

/**
 * Event handler function type
 */
export type EventHandler = (event: SecurityEvent) => void | Promise<void>;

/**
 * Event subscription handle
 */
export interface Subscription {
  /** Unsubscribe function */
  unsubscribe: () => void;
}

/**
 * EventBus for publishing and subscribing to security events
 */
export class EventBus {
  /** Event handlers by event type */
  private handlers: Map<EventType, Set<EventHandler>>;
  /** Global handlers for all events */
  private globalHandlers: Set<EventHandler>;
  /** Event history buffer */
  private history: SecurityEvent[];
  /** Maximum history size */
  private maxHistory: number;
  /** Dead letter queue for failed deliveries */
  private deadLetter: SecurityEvent[];
  /** Maximum dead letter size */
  private maxDeadLetter: number;

  /**
   * Create a new EventBus
   * @param maxHistory - Maximum events to keep in history
   * @param maxDeadLetter - Maximum failed events to keep
   */
  constructor(maxHistory: number = 1000, maxDeadLetter: number = 500) {
    this.handlers = new Map();
    this.globalHandlers = new Set();
    this.history = [];
    this.maxHistory = maxHistory;
    this.deadLetter = [];
    this.maxDeadLetter = maxDeadLetter;
  }

  /**
   * Subscribe to events of a specific type
   * @param type - Event type to subscribe to
   * @param handler - Event handler function
   * @returns Subscription handle
   * @example
   * ```typescript
   * const sub = eventBus.subscribe(EventType.AUTHENTICATION, (event) => {
   *   console.log('Auth event:', event.message);
   * });
   * ```
   */
  subscribe(type: EventType, handler: EventHandler): Subscription {
    if (!this.handlers.has(type)) {
      this.handlers.set(type, new Set());
    }
    this.handlers.get(type)!.add(handler);
    return {
      unsubscribe: () => {
        this.unsubscribe(type, handler);
      },
    };
  }

  /**
   * Subscribe to all events
   * @param handler - Event handler function
   * @returns Subscription handle
   * @example
   * ```typescript
   * eventBus.subscribeAll((event) => {
   *   logger.info('Event received', { type: event.type });
   * });
   * ```
   */
  subscribeAll(handler: EventHandler): Subscription {
    this.globalHandlers.add(handler);
    return {
      unsubscribe: () => {
        this.globalHandlers.delete(handler);
      },
    };
  }

  /**
   * Unsubscribe from events of a specific type
   * @param type - Event type
   * @param handler - Handler to remove
   * @example
   * ```typescript
   * eventBus.unsubscribe(EventType.AUTHENTICATION, handler);
   * ```
   */
  unsubscribe(type: EventType, handler: EventHandler): void {
    const typeHandlers = this.handlers.get(type);
    if (typeHandlers) {
      typeHandlers.delete(handler);
    }
  }

  /**
   * Publish an event to all subscribers
   * @param event - Event to publish
   * @example
   * ```typescript
   * eventBus.publish({
   *   id: crypto.randomUUID(),
   *   type: EventType.AUTHENTICATION,
   *   severity: EventSeverity.INFO,
   *   timestamp: new Date(),
   *   source: 'auth-service',
   *   message: 'User logged in',
   *   metadata: { userId: '123' },
   * });
   * ```
   */
  async publish(event: SecurityEvent): Promise<void> {
    this.addToHistory(event);

    const typeHandlers = this.handlers.get(event.type) || new Set();
    const allHandlers = [...this.globalHandlers, ...typeHandlers];

    for (const handler of allHandlers) {
      try {
        await handler(event);
      } catch (error) {
        this.deadLetter.push(event);
        if (this.deadLetter.length > this.maxDeadLetter) {
          this.deadLetter.shift();
        }
      }
    }
  }

  /**
   * Get event history
   * @param type - Optional filter by event type
   * @param limit - Maximum number of events to return
   * @returns Array of security events
   * @example
   * ```typescript
   * const authEvents = eventBus.getHistory(EventType.AUTHENTICATION, 10);
   * ```
   */
  getHistory(type?: EventType, limit: number = 100): SecurityEvent[] {
    let events = this.history;
    if (type) {
      events = events.filter((e) => e.type === type);
    }
    return events.slice(-limit);
  }

  /**
   * Get dead letter queue (failed event deliveries)
   * @param limit - Maximum number of events to return
   * @returns Array of failed events
   * @example
   * ```typescript
   * const failed = eventBus.getDeadLetter();
   * ```
   */
  getDeadLetter(limit: number = 100): SecurityEvent[] {
    return this.deadLetter.slice(-limit);
  }

  /**
   * Get the number of subscribers for an event type
   * @param type - Event type
   * @returns Subscriber count
   */
  getSubscriberCount(type: EventType): number {
    return (this.handlers.get(type)?.size || 0) + this.globalHandlers.size;
  }

  /**
   * Clear event history
   */
  clearHistory(): void {
    this.history = [];
  }

  /**
   * Clear dead letter queue
   */
  clearDeadLetter(): void {
    this.deadLetter = [];
  }

  /**
   * Add event to history buffer
   * @param event - Event to add
   */
  private addToHistory(event: SecurityEvent): void {
    this.history.push(event);
    if (this.history.length > this.maxHistory) {
      this.history.shift();
    }
  }
}

let _eventBus: EventBus | null = null;

/**
 * Get the global event bus singleton
 * @param maxHistory - Max history size (only on first call)
 * @param maxDeadLetter - Max dead letter size (only on first call)
 * @returns EventBus instance
 * @example
 * ```typescript
   const eventBus = getEventBus();
   * eventBus.publish(myEvent);
   * ```
   */
export function getEventBus(maxHistory?: number, maxDeadLetter?: number): EventBus {
  if (!_eventBus) {
    _eventBus = new EventBus(maxHistory, maxDeadLetter);
  }
  return _eventBus;
}
